test_data/asm/wrap_gl.asm

```diff
@@ -0,0 +1,139 @@
+// Wraps an arbitrary field element into a u32, on the Goldilocks field.
+machine WrapGL(RESET, operation_id) {
+
+    operation wrap<0> in_acc -> output;
+
+    constraints {
+        macro is_nonzero(X) {
+            match X {
+                0 => 0,
+                _ => 1,
+            }
+        };
+        macro is_zero(X) { 1 - is_nonzero(X) };
+
+        col fixed BYTES(i) { i % 256 };
+
+        // Latch and operation ID
+        col fixed RESET(i) { is_zero((i % 8) - 7) };
+        col witness operation_id;
+
+        // 1. Decompose the input into bytes
+
+        // The byte decomposition of the input, in little-endian order
+        // and shifted forward by one (to use the last row of the
+        // previous block)
+        col witness bytes;
+        // Puts the bytes together to form the input
+        col witness in_acc;
+        // Factors to multiply the bytes with
+        col fixed FACTOR(i) { 1 << (((i + 1) % 8) * 8) };
+
+        {bytes} in {BYTES};
+        in_acc' = in_acc * (1 - RESET) + bytes * FACTOR;
+
+        // 2. Build the output, packing the least significant 4 byte into
+        //    a field element
+        col witness output;
+        col fixed FACTOR_OUTPUT = [0x100, 0x10000, 0x1000000, 0, 0, 0, 0, 1]*;
+        output' = output * (1 - RESET) + bytes * FACTOR_OUTPUT;
+
+        // 3. Check that the byte decomposition does not overflow
+        //
+        //    Skipping this step would work but it wouldn't be sound, because
+        //    the 8-byte decomposition could overflow, since the Goldilocks
+        //    prime 2**64 - 2**32 + 1 is smaller than 2^64.
+        //
+        //    The approach is to compare the byte decomposition with that of
+        //    the maximum possible value (0xffffffff00000000) byte by byte,
+        //    from most significant to least significant (i.e., going backwards).
+        //    A byte can only be larger than that of the max value if any previous
+        //    byte has been smaller.
+
+        // This is an example for input 0xfffffffeffffffff:
+        // Row     RESET   bytes   BYTES_MAX  lt      was_lt  gt
+        // -1      0x1     0xff    0x0        0x0     0x1     0x1
+        //  0      0x0     0xff    0x0        0x0     0x1     0x1
+        //  1      0x0     0xff    0x0        0x0     0x1     0x1
+        //  2      0x0     0xff    0x0        0x0     0x1     0x1
+        //  3      0x0     0xfe    0xff       0x1     0x1     0x0  # 0xfe < 0xff, so now greater bytes are allowed
+        //  4      0x0     0xff    0xff       0x0     0x0     0x0
+        //  5      0x0     0xff    0xff       0x0     0x0     0x0
+        //  6      0x0     0xff    0xff       0x0     0x0     0x0
+        //  7      0x1     ----    ----       ---     ---     ---
+
+        // Bytes of the maximum value, in little endian order, rotated by one
```

Comment:
what's "rotated by one", and why?

Reply:
It's a performance optimization, similar to how `block_lookup_or` works: On the latch row, only the input and output has to be available, all columns can already be used by the next block.